### ISSUE TITLE
Consider making "messages" and therefore "message generators" optional to all operation APIs

### ISSUE THREAD

**Issue by tplooker** [decentralized-identity/bbs-signature#117]
Now we have introduced the concept of an optional header in #95 I want to consider whether we allow for constructions of BBS signatures which do not make use of the "selective disclosure" feature, instead just use BBS as an ordinary group signature. I believe there are several interesting usecases such as privacy preserving access tokens or hardware attestations that would likely use BBS in this form.

In short my proposal is for both the header and messages (along with message generators) to be optional. Meaning a signer could create a signature that has neither, either or both a header and array of messages.

**Comment by BasileiosKal** [comment]
+1. In this case, the issuer will just sign their public key and i dont really see any issues with that. We can specify that L >= 0 (where L the number of messages that can be selectively disclosed).

**Comment by tplooker** [comment]
One thing to check is in making this change whether there is anything different in the SpkGen or SpkVerify operations?

**Comment by BasileiosKal** [comment]
I don’t think there is. There is nothing special about a signature on just 1 message (the domain which will be just the hash of the signers PK)

**Comment by tplooker** [comment]
One other thing to note is that because the messages and message generators are seperate arguments past to the sign verify etc functions by making these arguments optional we create the case where only one could be supplied which would be invalid, for example just messages and no generators or vice versa.

**Comment by BasileiosKal** [comment]
That’s a good point. I think a check must be added to ensure that the number of messages is the same with the number of generators (regardless of what we decide here). IMO this will also resolve this issue.

**Comment by andrewwhitehead** [comment]
The method could also accept an optional list of (generator, message) pairs.

**Comment by BasileiosKal** [comment]
> The method could also accept an optional list of (generator, message) pairs.

We could but IMO it may be hard to get the values out of that list of tuples to be used in the operation, i.e., if the input is `[(H_1, msg_1), (H_2, msg_2), ...., (H_n, msg_n)]` wouldn't it be a little unclear how we will get the generators out to be passed in the `domain` etc.?? 

Further i like passing a list of messages because it makes the API as similar as possible to standard "known" digital signatures, so people can draw clear lines between them (i.e., say stuff like "it's just a digital signature but instead of 1 message you will pass a list of messages" etc), just to help with adoption.

That said, i do recognize that the pairs approach makes sense and I'm definitely not opposed to it. I would like to see a PR to get a better idea of how it is going to look.

**Comment by tplooker** [comment]
The other option I would like us to consider is making the message generators parameters instead of inputs. How these generators are used most commonly in a cipher suite I suspect will be the determining factor in whether we make this change. IMO having not seen any security issues with sharing generators across signatures and issuers I think implementations are more likely to treat message generators as globally shared values, cached by implementations. If we believe this is going to be normal then documenting the message generators as parameters rather than inputs makes more sense to me.

**Comment by BasileiosKal** [comment]
So to recap it seems that there are three options,

1. The current approach i.e., a messages list and a generator list as input.  This will look like the following
     ```
    signature = Sign(SK, PK, (msg_1, ..., msg_L), (H_1, ..., H_L))
     ```
2. @tplooker's proposal to pass the generators as parameters. This will look like,
    ````
    signature = Sign(SK, PK, (msg_1, ..., msg_L))
   
    Parameters: 
    - (H_1, ..., H_L): The generators used for the signature
    ````
3. @andrewwhitehead's proposal of passing one list with (generator, message) pairs. This will look like,
    ```
    signature = Sign(SK, PK, [(H_1, msg_1), ..., (H_L, msg_L)])
    ```

My thoughts (see also the discussion above for more context): 
- Option 2 does give the clearer API. I'm only worried that we may break the separation between what constitutes as input and what as parameter.
- Option 3 does remove the need to have to check if the messages and the generators lists have the same length. It may create the need to also pass an additional list of generators to ProofVerify though because that operation requires all the generators, not just the ones corresponding to the revealed messages. This may create some inconsistency between the ProofGen and ProofVerify api?? (not really sure about that)

Currently im leaning more towards option 2, if we can sufficiently justify the choice to the cfrg.

**Comment by tplooker** [comment]
Labeling as pending close ahead of WG call on the 4th of july

**Comment by tplooker** [comment]
Discussed on 4th of july WG call, closing as resolved via #183
